vowpalwabbit/conditional_contextual_bandit.cc

```diff
@@ -1,23 +1,147 @@
 #include "conditional_contextual_bandit.h"
-
+#include "reductions.h"
 #include "example.h"
 #include "global_data.h"
 #include "cache.h"
 
 #include <numeric>
 #include <algorithm>
 
+using namespace LEARNER;
 using namespace VW;
+using namespace VW::config;
+
+template <bool is_learn>
+void do_actual_learning(CCB::ccb& data, multi_learner& base, multi_ex& examples) {
```

Comment:
Consider renaming to `learn_or_predict`. `do_actual_learning` isn't standard inside vw

Reply:
do_actual_learning was a relic prior to @rajan-chari refactoring to support multiline examples directly.  In essence, there was a fake learn_or_predict that gathered examples, dispatching them to do_actual_learning when the full multiline example was created.  Nowadays, that curmudgeon is not needed.